master/buildbot/data/resultspec.py

```diff
@@ -325,18 +333,28 @@ def includeFields(d):
 
             # precompute the ordering functions and sort
             if self.order:
-                order = [(lambda a, b, k=k[1:]: nonecmp(b[k], a[k]))
-                         if k[0] == '-' else
-                         (lambda a, b, k=k: nonecmp(a[k], b[k]))
-                         for k in self.order]
-
-                def cmpFunc(a, b):
-                    for f in order:
-                        c = f(a, b)
-                        if c:
-                            return c
-                    return 0
-                data.sort(cmp=cmpFunc)
+                # Do a multi-level sort by sorting from
+                # the least significant key to the most
+                # significant key.  If we are given the keys
+                # ('lastName', 'firstName', 'age'):
+                #   - first sort by 'age'
+                #   - then sort by 'firstName'
+                #   - then sort by 'lastName'
+                for k in reversed(self.order):
+                    doReverse = False
+                    if k[0] == '-':
+                        # If we get a key '-lastName',
+                        # it means sort by 'lastName' in reverse.
+                        k = k[1:]
+                        doReverse = True
+
+                    def keyFunc(a, sortBy=k):
+                        val = a[sortBy]
+                        if val is None:
+                            val = MinType()
+                        return val
+
+                    data.sort(key=keyFunc, reverse=doReverse)
```

Comment:
we are doing N sorts instead of one? sounds like a big performance regression. This sorting is in the heart of our data api, so we need to make it optimized (at least bigO wise, no need for micro optimization).


http://stackoverflow.com/questions/5212870/sorting-a-python-list-by-two-criteria

second answer looks like the one we need:

    sorted(list, key=lambda x: (x[0], -x[1]))

We need to build a function that will turn 

    ('lastName', '-firstName', 'age')

to

    lambda x: (NoneAndReverseComparator(x['lastname'], reverse=False), (NoneAndReverseComparator(x['firstName'], reverse=True), NoneAndReverseComparator(x['age'], reverse=False))

with NoneAndReverseComparator to implement __eq__ and __lt__ taking in account None to be the minimum, and optionally reverse the output of __lt__.